### DISCUSSION TITLE
Wrong calculation of "Power House Load Yield"

### DISCUSSION THREAD

**Discussion by cmovetus75** [JensenNick/huawei_solar_pees#68] [Question / Answer]
After installing the package with the PEES sensors everything has worked fine except the calculation of "Power House Load Yield" (sensor.power_house_load) and all sensors that further use this value and associated utility meters under some circumstances.
In particular, in forced charge and forced discharge modes of use of the battery, the formula used in the template sensor to calculate the load of the house that is being sourced from PV Panels (Power House Load Yield) does not cover properly this use case.

The current pseudo formula is more or less :
**If** (House Load > PV Yield) **then** House Load Yield = PV Yield
**Else** House Load Yield = House Load

If in forced charge mode, all PV Yield is sent to the Battery (for later use when the electricity cost is more expensive), but with the above formula would assume that the PV Yield is being sent to the House Load.

I have being using for some hours a new formulation that seems to be working well. I have identified 4 different scenarios, that require a diferent formula according to the power flow diagram:

**If** (batteries_charge_discharge_power > 0) (charging)
	If PV Yield > Charge -> House Load Yield = House Load – Import 
	If PV Yield ≤ Charge -> House Load Yield = 0
**Else** (discharging)
	If Discharge < Load -> House Load Yield = PV Yield – Export 
	If Discharge ≥ Load -> House Load Yield = 0 
I have attached the modified PEES yaml definition that I am currently using.

[huawei_solar_pees (Power House Load Yield corrected).txt](https://github.com/user-attachments/files/17006538/huawei_solar_pees.Power.House.Load.Yield.corrected.txt)

**Comment by JensenNick** [comment]
I remember playing around with the formulas when Forced Charge first was introduced. My conclusion back then was to leave things as they were. What do you base your assumption on - that all PV Yield is sent to the battery in Forced Charge?

It would make sense from a "efficiency" POV if the system behaves like you describe.

But if you make the changes you suggest you increase the house load consumption from grid and improve the battery economy - I think it is debatable if that is the "correct" way. I'll have a look at it.

**Reply by cmovetus75** [reply]
In "Forced Charge" mode you set a max power to charge (ex. 2500W) and a duration of the forced charge period (ex. 120min). So, lets imagine that the PV panels are producing 4000W, then, during the Forced Charge period, 2500W will be sent to the battery for charging and 1500W are available to either the house load or to export (if the house load is less than 1500W).
Conversely, if the panels are producing 1000W, then the battery is only charging at 1000W rate (if, as in my case you have disabled the charge from grid option) and all the production from the PV panels is sent to the battery. The house load is being fully powered by the grid in that conditions. In this case you could say that you are (in that precise snapshot) incrising the load consumption from the grid, but if you tune the timing properly, at the end of the day, the energy consumed from the grid would be the same as if you had beeing feeding the PV yield directly to the house... with the main difference that the use of the solar energy would be in the most economically interesting periods. In theory it improves battery economy as a result (minor corrections should be taken into account to be more precise, like the potential increase in degradation because of increase of cycling of the battery or the loss of efficiency in the roundtrip charge-discharge instead of direct PV yield consumption)

**Reply by JensenNick** [reply]
Your approach may be right when only looking at the energy consumption, but not when it comes to the financial part of the calculations. I, like many others, want to know how much the battery contributes financially, including when it is paid back. This is done by comparing the two situations with and without battery.

By shifting the distribution/consumption as you do, you distort the financial calculations. You get an artificially high household consumption in the middle of the day (when force charging) that you otherwise would not have had if you did not have a battery and you are charging the battery with PV production, that would have been used to cover the household consumption and be sold if you didn't have a battery. You're artificially improving the battery economy.

I don't disagree that your approach could improve battery economy even more, but we would loose the ability to compare the two situation (with and without battery) which I would prioritize.

As mentioned I did play around with forced charging and also did test shifting my consumption. But in the middle there was a SW update from Huawei which prevented to send all production to the battery while there was a house load. It sounds like they have changed that back again. This is again another reason I'm reluctant to make the changes you propose.
